boards/nordic/nrf52840dk/src/main.rs

```diff
@@ -914,8 +916,14 @@ pub unsafe fn start() -> (
     let _ = platform.pconsole.start();
     base_peripherals.adc.calibrate();
 
+    io::DEBUG_INFO.with(|debug_info| debug_info.put(io::DebugInfo {
+	chip,
+	processes,
+	process_printer,
+    }));
```

Comment:
This passes references to these to a potential future panic "thread" while other references to them live on and continue to be used.

Do we have a good sense for whether this is safe? I agree that the old approach is probably unsafe, so this is not necessarily making things worse. But it is not obvious to me that this is safe, and if it is not I sort of think we shouldn't make it look like it may be.

Reply:
The fundamental danger with this infrastructure is that we're providing a reference of `T: ?Sync` in a `Sync` wrapper, and I hardly think that `chip` or `processes` are `Sync`. So strictly speaking this is probably not sound as would be required by Rust.

This is a hard problem to solve, and while the introduction of `CoreLocal` does not change our premise and actual safety guarantees here, it's properties around `Sync` are what's problematic... In fact, I never thought of panics as being virtually a different thread (which they definitely are!), but only ISRs.

I think that this here is symptomatic of a larger body of unsoundness in Tock when it comes to reasoning about which state can be assumed to be valid in these non-single-thread-synchronous circumstances. A panic may occur precisely because an invariant in some struct reachable through those reference is violated, and Rust doesn't give us particularly many guarantees around their validity beyond that point. So, it's really not only that `chip` and `processes` are `!Sync`, but that they may not be valid at all in a panic context. An unergonomic, expensive and ugly fix is to move all state required in the panic handler into some "atomic" statics (perhaps wrapped in a `CoreLocal`), that are `Sync`, have a `Cell`-like interface, can only wrap datatypes that are guaranteed to be updated atomically w.r.t. a single hardware thread, and are unconditionally valid.

In short, I don't think we can or should fix this in this PR.

Reply:
Panic is not in a different thread

Reply:
> I hardly think that chip or processes are Sync

Why? Multiple points of entry does not mean they are multiple threads of execution. They are not multiple threads of execution and that is pretty important to the Tock kernel functioning correctly generally.

There are exactly two kinds of software things that run concurrently with the kernel:

- Actual Processes, which do not have access to kernel memory
- ISRs which are specifically designed to do almost nothing except switch back to the kernel thread and retain the interrupt flag. They really _should not_ touch any memory shared with the broader kernel.

Everything else is single threaded. If it is not single threaded, it is a bug.

Reply:
> I hardly think that chip or processes are Sync

Aside from your actual comment text which I agree with, the aforementioned variables `chip` and `process` won't be sync because at least, as far as I know, they all contain some form of `core::cell::UnsafeCell` that is `!Sync`.

Reply:
Oh... I retract! You're of course right. I totally misunderstood that you were referring to the variables rather than the crates/modules. Sorry!